content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/vmware_best_fit_least_utilized.rb

```diff
@@ -1,37 +1,74 @@
 #
-$evm.log("info", "Args:    #{MIQ_ARGS.inspect}")
-
-# Get variables
-prov = $evm.root["miq_provision"]
-vm = prov.vm_template
-raise "VM not specified" if vm.nil?
-ems = vm.ext_management_system
-raise "EMS not found for VM [#{vm.name}]" if ems.nil?
-
-# Log space required
-$evm.log("info", "vm=[#{vm.name}], space required=[#{vm.provisioned_storage}]")
-
-host = storage = nil
-min_registered_vms = nil
-prov.eligible_hosts.each do |h|
-  next if h.maintenance
-  next unless h.power_state == "on"
-  nvms = h.vms.length
-  if min_registered_vms.nil? || nvms < min_registered_vms
-    storages = h.writable_storages.find_all { |s| s.free_space > vm.provisioned_storage } # Filter out storages that do not have enough free space for the Vm
-    s = storages.max_by(&:free_space)
-    unless s.nil?
-      host    = h
-      storage = s
-      min_registered_vms = nvms
+module ManageIQ
+  module Automate
+    module Infrastructure
+      module VM
+        module Provisioning
+          module Placement
+            class VmwareBestFitLeastUtilized
+              def initialize(handle = $evm)
+                @handle = handle
+              end
+
+              def main
+                @handle.log("info", "vm=[#{vm.name}], space required=[#{vm.provisioned_storage}]")
+                best_fit_least_utilized
+              end
+
+              private
+
+              def request
+                @request ||= @handle.root["miq_provision"].tap do |req|
+                  log_and_raise('miq_provision object not passed in') if req.nil?
+                end
+              end
+
+              def vm
+                @vm ||= request.vm_template.tap do |vm|
+                  log_and_raise('VM not specified') if vm.nil?
+                end
+              end
+
+              def storage_profile_id
+                @storage_profile_id ||= request.get_option(:placement_storage_profile).tap do |sp|
+                  @handle.log("info", "Selected storage_profile_id: #{sp}") if sp
+                end
+              end
+
+              def best_fit_least_utilized
+                host = storage = min_registered_vms = nil
+                request.eligible_hosts.select { |h| !h.maintenance && h.power_state == "on" }.each do |h|
+                  next if min_registered_vms && h.vms.size >= min_registered_vms
+
+                  storages = h.writable_storages.find_all { |s| s.free_space > vm.provisioned_storage } # Filter out storages that do not have enough free space for the Vm
+                  storages.select! { |s| s.storage_profiles.pluck(:id).include?(storage_profile_id) } if storage_profile_id
+
+                  s = storages.max_by(&:free_space)
+                  next if s.nil?
+                  host    = h
+                  storage = s
+                  min_registered_vms = h.vms.size
+                end
+
+                # Set host and storage
+                request.set_host(host) if host
```

Comment:
@lfu What happens if host and storage are both nil we can't find any appropriate host or storage should we record that as an error and stop the process

Reply:
The backend raises an error if the required resources have not been selected.  The best-fit methods do not have to deal with it.

See https://github.com/ManageIQ/manageiq-providers-vmware/blob/master/app/models/manageiq/providers/vmware/infra_manager/provision/placement.rb#L13-L15

Reply:
@gmcculloug Should there be a warning message in this method if we can't pick a host or a storage?

Reply:
That's fine if you want to add one.  The line immediately below this also logs if we have either object: `@handle.log("info", "vm=[#{vm.name}] host=[#{host}] storage=[#{storage}]")`

Reply:
I don't think adding another log message should be necessary as there is already this line:
```
@handle.log("info", "vm=[#{vm.name}] host=[#{host}] storage=[#{storage}]")
```
at the end of this method, then later https://github.com/ManageIQ/manageiq-providers-vmware/blob/master/app/models/manageiq/providers/vmware/infra_manager/provision/placement.rb#L13-L15 as @gmcculloug has pointed out.

Reply:
@lfu @gmcculloug 
If there is no host or storage that message should atleast become a warning otherwise it will get lost in the log, if its a warning at least it would stand out. Can we change the level to warning if host or storage is nil

Reply:
An error message would be thrown out if host or storage is nil.

```
    raise MiqException::MiqProvisionError, "Destination placement_ds_name not provided" if datastore.nil?
    raise MiqException::MiqProvisionError, "Destination placement_host_name and placement_cluster_name not provided" if host.nil? && cluster.nil?
```

Reply:
@lfu 
That error message happens in a completely disconnected flow when the provision starts. The fact that the placement method didn't find a suitable host or storage is never reported and lost in a log message where the value is nil. As far as debugging is concerned we wont know that this method didn't find a suitable host or storage. Its important to report an error/warning when it first happens rather than have some other process down the road throw an error that it didn't get the required value.

Reply:
@mkanoor I would like to suggest that we address this issue in a separate PR or git issue as not to hold up the refactoring and storage profile fix this PR is trying to address.

But I wanted to note here, in my mind we have always allowed the backend to determine if all the requirements were met for the task, it did not fall on the automate script to know all the details of what the provider required.  And a single script may not provide all mandatory fields, they could be provided through another means.

For example, a user could use this script and potentially follow it with another placement script, as not to change our default methods, that uses different logic to provide the missing information.  Or the backend implementation could potentially change to determine missing fields differently based on new features or settings.

Having the automate method know these details does not see like the best place.

Again, I'm not against discussing this, but it feels like a different topic and probably should not hold up reviewing/merging this PR.